server: add verbose logging when graceful drain range lease transfer stalls

cameronnunez opened the pull request:
Informs #65659.

Debugging graceful drain is challenging due to limited information. It is not clear what
is going wrong from the logs. This patch adds more detailed logging when graceful drain
range lease transfer encounters issues. Specifically, range lease transfer attempt
information (when an attempt occurs, duration blocked on transfer attempt, and the lease
that is being transferred) is logged.

The reason behind the failure of the transfer attempt may not be logged on the
draining node. This is because this information lies on the node that serves the draining
request. Currently, we do not have logging that aggregates all of this information
in one place. See #74158.

Release note (cli change): If graceful drain range lease transfer encounters issues,
verbose logging is automatically set to help with troubleshooting.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68488)
<!-- Reviewable:end -->

knz reviewed (commented):
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cameronnunez)

___
*[pkg/cli/start.go, line 826 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgSHhulB4gze4X6jBOt:-MgSHhulB4gze4X6jBOu:bumnxtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e4da50af1d444d63bb3df03e04d92d478a7a89/pkg/cli/start.go#L826)):*
> ```Go
> 			// via a 2nd signal.
> 			for {
> 				var prevRemaining = remaining
> ```

You need to shift the var outside the for loop so that it retains its value between iterations
___
*[pkg/cli/start.go, line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgSHgNS0mIw15hfEcLL:-MgSHgNT686QUmASb3U7:beszife) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e4da50af1d444d63bb3df03e04d92d478a7a89/pkg/cli/start.go#L832)):*
> ```Go
> 					break
> 				}
> 				if remaining == prevRemaining {
> ```

Can you make this an inequality - we also want to trigger verbose mode if the remaining work increases


<!-- Sent from Reviewable.io -->

cameronnunez reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/start.go, line 826 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgSHhulB4gze4X6jBOt:-MghOD-CCuIPvnZ2fQKN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e4da50af1d444d63bb3df03e04d92d478a7a89/pkg/cli/start.go#L826)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

You need to shift the var outside the for loop so that it retains its value between iterations
</blockquote></details>

Done.
___
*[pkg/cli/start.go, line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgSHgNS0mIw15hfEcLL:-MghODYODTixmFXwIdQR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68e4da50af1d444d63bb3df03e04d92d478a7a89/pkg/cli/start.go#L832)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you make this an inequality - we also want to trigger verbose mode if the remaining work increases
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cameronnunez)

___
*[pkg/cli/start.go, line 829 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-Mgpfq6h-9YgLirj_qvx:-Mgpfq6h-9YgLirj_qvy:bixza2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/cli/start.go#L829)):*
> ```Go
> 			// to help with troubleshooting.
> 			var (
> 				remaining     uint64
> ```

missing `var verbose bool`
___
*[pkg/kv/kvserver/store.go, line 1149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgpfIoS0InjGvQWaoxg:-MgpfIoS0InjGvQWaoxh:bku4kms) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/kv/kvserver/store.go#L1149)):*
> ```Go
> 					)
> 					if transferStatus != transferOK {
> 						format := "failed to transfer lease %s for range %s when draining: %s"
> ```

make this a `const`

also use `%v` at the end not %s
___
*[pkg/kv/kvserver/store.go, line 1150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgpfK0p1ls8L7CEHNvn:-MgpfK0p1ls8L7CEHNvo:b6mp0yr) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/kv/kvserver/store.go#L1150)):*
> ```Go
> 					if transferStatus != transferOK {
> 						format := "failed to transfer lease %s for range %s when draining: %s"
> 						var outcome string
> ```

`var outcome interface{}`
___
*[pkg/kv/kvserver/store.go, line 1152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgpfLu79Qdtc2JvnXGM:-MgpfLu79Qdtc2JvnXGN:b9p5p0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/kv/kvserver/store.go#L1152)):*
> ```Go
> 						var outcome string
> 						if err != nil {
> 							outcome = err.Error()
> ```

`outcome = err`
___
*[pkg/kv/kvserver/store.go, line 1154 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgpfNn8AzBX4oWMYzGK:-MgpfNn94pS0omIQvTNI:b-kqey6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/kv/kvserver/store.go#L1154)):*
> ```Go
> 							outcome = err.Error()
> 						} else {
> 							outcome = transferStatus.String()
> ```

`outcome = transferStatus`


<!-- Sent from Reviewable.io -->

cameronnunez reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/start.go, line 829 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-Mgpfq6h-9YgLirj_qvx:-MgrFElX7hXK-JiWwYPE:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/cli/start.go#L829)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

missing `var verbose bool`
</blockquote></details>

Added.
___
*[pkg/kv/kvserver/store.go, line 1149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgpfIoS0InjGvQWaoxg:-MgrFGcy28RxOB7BRw4_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/kv/kvserver/store.go#L1149)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

make this a `const`

also use `%v` at the end not %s
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store.go, line 1150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgpfK0p1ls8L7CEHNvn:-MgrFHTq0qxMfxWIbCGs:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/kv/kvserver/store.go#L1150)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`var outcome interface{}`
</blockquote></details>

Fixed.
___
*[pkg/kv/kvserver/store.go, line 1152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgpfLu79Qdtc2JvnXGM:-MgrFLyW1uJTV4j067Km:b-4hvi0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/kv/kvserver/store.go#L1152)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`outcome = err`
</blockquote></details>

Got it.
___
*[pkg/kv/kvserver/store.go, line 1154 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MgpfNn8AzBX4oWMYzGK:-MgrFO2XDDurkQFCgZHI:b-4hvi0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ed0e2e6f0691bc684d7f9837e17dfb17a60734/pkg/kv/kvserver/store.go#L1154)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`outcome = transferStatus`
</blockquote></details>

Got it.


<!-- Sent from Reviewable.io -->

rauchenstein reviewed (commented):
> Reviewed 11 of 12 files at r4, 2 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @cameronnunez, @knz, and @rauchenstein)

___
*[pkg/cli/quit.go, line 122 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MqzpNam910hajPhxeJh:-MqzpNam910hajPhxeJi:b-qm3jca) ([raw file](https://github.com/cockroachdb/cockroach/blob/fab55d4617ef064f405ae262226be3595ba77a93/pkg/cli/quit.go#L122)):*
> ```Go
> }
> 
> func doDrainNoTimeout(ctx context.Context, c serverpb.AdminClient) (hardError, _ bool, err error) {
> ```

Returning an unnamed and undocumented bool.  As is, the only way to figure out what this represents is to look at the implementation.
___
*[pkg/cli/quit.go, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MqzuCCd1OMk571JUZCG:-MqzuCCe8FJuQF0YgZDG:b3of3hr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fab55d4617ef064f405ae262226be3595ba77a93/pkg/cli/quit.go#L131)):*
> ```Go
> 
> 	var (
> 		remaining     = ^uint64(0)
> ```

math.MaxUint64 may be more readable for those who don't like thinking about bitwise representations.
___
*[pkg/cli/quit.go, line 203 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-Mqzr1zA3tMv0qkqXIxK:-Mqzr1zA3tMv0qkqXIxL:bb3ttf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fab55d4617ef064f405ae262226be3595ba77a93/pkg/cli/quit.go#L203)):*
> ```Go
> 		// If range lease transfer stalls or the number of remaining leases
> 		// somehow increases, verbosity is set to help with troubleshooting.
> 		if !verbose && remaining >= prevRemaining {
> ```

nit: !verbose check is unnecessary.


<!-- Sent from Reviewable.io -->

cameronnunez reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @knz, and @rauchenstein)

___
*[pkg/cli/quit.go, line 122 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MqzpNam910hajPhxeJh:-MrNmZiO47Z2NUx1j2n-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fab55d4617ef064f405ae262226be3595ba77a93/pkg/cli/quit.go#L122)):*
<details><summary><i>Previously, rauchenstein wrote…</i></summary><blockquote>

Returning an unnamed and undocumented bool.  As is, the only way to figure out what this represents is to look at the implementation.
</blockquote></details>

Done.
___
*[pkg/cli/quit.go, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MqzuCCd1OMk571JUZCG:-MrNmZyW4NihnBJKHAgv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fab55d4617ef064f405ae262226be3595ba77a93/pkg/cli/quit.go#L131)):*
<details><summary><i>Previously, rauchenstein wrote…</i></summary><blockquote>

math.MaxUint64 may be more readable for those who don't like thinking about bitwise representations.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz commented:
related: #74158.

rauchenstein reviewed (commented):
LGTM, though I'd rather someone with more roachtest familiar pass off on that part.

> Reviewed 10 of 12 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-:-MsfpeaG0irmK8yNTvr-:b-eah8f8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @knz, and @rauchenstein)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Didn't review the actual logging changes, but it looks like Raphael has us covered there. Focused on the roachtest.

> Reviewed 1 of 12 files at r4, 8 of 12 files at r9, 2 of 3 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-:-MtDTG5K5E95sl7jmMNT:bgfm66g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @cameronnunez, @knz, and @rauchenstein)

___
*[pkg/cmd/roachtest/tests/acceptance.go, line 71 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDS3jkEH7mMHUOmH2N:-MtDS3jkEH7mMHUOmH2O:bz2p2f0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/acceptance.go#L71)):*
> ```Go
> 			{name: "cluster-init", fn: runClusterInit},
> 			{name: "rapid-restart", fn: runRapidRestart},
> 			{name: "slow-drain", fn: runSlowDrain, numNodes: 6, timeout: 30 * time.Minute},
> ```

I'm hesitant about adding this as an acceptance test - these run on *every* CI. Is that really what you want? I figured you'd add this as a free-standing test, since in the test it seems like you're running it for tens of minutes.
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 40 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDSM-a7ck9qBs9QUzk:-MtDSM-a7ck9qBs9QUzl:burkn7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L40)):*
> ```Go
> 	var verboseStoreLogRe = regexp.MustCompile("failed to transfer lease")
> 
> 	if err := c.PutE(ctx, t.L(), t.Cockroach(), "./cockroach", c.All()); err != nil {
> ```

nit: unclear why this isn't just the shorter `c.Put`.
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 85 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDSi9s25pd9vk7_aOQ:-MtDSi9s25pd9vk7_aOR:bob9kqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L85)):*
> ```Go
> 
> 	var m *errgroup.Group
> 	dur := 30 * time.Minute
> ```

How can this work? I saw earlier that you're giving this test a 30 minute timeout only.
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 86 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDT1vZCS3MZ0NMl02R:-MtDT1vZCS3MZ0NMl02S:b-at4n2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L86)):*
> ```Go
> 	var m *errgroup.Group
> 	dur := 30 * time.Minute
> 	m, ctx = errgroup.WithContext(ctx)
> ```

Please use a `c.NewMonitor`, also you never `.Wait()` on this errgroup, so what's the point?
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 91 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDSwTtExlBa-WhxZ1a:-MtDSwTtExlBa-WhxZ1b:bhmy5mt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L91)):*
> ```Go
> 			return c.RunE(ctx, c.Node(pinnedNodeID),
> 				fmt.Sprintf("./cockroach workload run kv --max-rate 500 --tolerate-errors --duration=%s {pgurl:1-%d}",
> 					dur.String(), numNodes-4,
> ```

The numNodes-4 is confusing to me, mind adding a comment?
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 101 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDT6PF7RLQtWqdGrie:-MtDT6PF7RLQtWqdGrif:b-k2e38) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L101)):*
> ```Go
> 
> 	// Drain 5 nodes from the cluster, resulting in immovable leases.
> 	m, ctx = errgroup.WithContext(ctx)
> ```

Ditto about monitor


<!-- Sent from Reviewable.io -->

cameronnunez committed:
server: add verbose logging when graceful drain range lease transfer stalls

Debugging graceful drain is challenging due to limited information. It is not clear what
is going wrong from the logs. This patch adds more detailed logging when graceful drain
range lease transfer encounters issues. Specifically, range lease transfer attempt
information (when an attempt occurs, duration blocked on transfer attempt, and the lease
that is being transferred) is logged.

The reason behind the failure of the transfer attempt may not be logged on the
draining node. This is because this information lies on the node that serves the draining
request. Currently, we do not have logging that aggregates all of this information
in one place. See #74158.

Release note (cli change): If graceful drain range lease transfer encounters issues,
verbose logging is automatically set to help with troubleshooting.

knz reviewed (commented):
cli and server changes LGTM.
A question about the roachtest.

The commit message (as reported by reviewable) is weird.

When you edit that commit message, please explain in there (not in the release note) that this only covers half of the story, because this logging only tells us whether transfers are attempted or not, and they don't tell us why they fail. You can then go on to explain that the precise cause is decided on _another_ node, the one that's driving the transfer (that's not the same node as the one that's decommissioning), and we don't yet have logging/tracing infra to gather that data in one place properly.

I recall there was a discussion with Andrei about this somewhere? Can you find it? And then link a URL in the commit message.

> Reviewed 4 of 12 files at r9, 9 of 9 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-:-MuBgeWV0EYX2HJXeE5G:b9ajfd0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @cameronnunez, and @rauchenstein)

___
*[-- commits, line 2 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MuBgEiA9tNFqTcUIvzO):*
What's up with this commit message? I think it's wrong and belongs to a different PR.
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 132 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MuBgYz88weJCUGFGExe:-MuBgYz88weJCUGFGExf:bhrbv5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/81379d2270a553947308d4a2bc680ca506ed04be/pkg/cmd/roachtest/tests/slow_drain.go#L132)):*
> ```Go
> 	for nodeID := 2; nodeID <= numNodes; nodeID++ {
> 		if err := c.RunE(ctx, c.Node(nodeID),
> 			fmt.Sprintf("grep -q '%s' logs/cockroach.log", verboseStoreLogRe),
> ```

I feel stupid, I'm not seeing it: can you point me to where `verboseStoreLogRe` is defined?
___
*[pkg/server/serverpb/admin.proto, line 408 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MuBfnBL4E_wvkN90Ndp:-MuBfnBM2u_8II_Imels:b678388) ([raw file](https://github.com/cockroachdb/cockroach/blob/81379d2270a553947308d4a2bc680ca506ed04be/pkg/server/serverpb/admin.proto#L408)):*
> ```protobuf
>   // in subsequent versions.
>   string node_id = 5;
>   // When true, more detailed information is logged during the drain phase.
> ```

nit: "during the range lease drain phase."


<!-- Sent from Reviewable.io -->

cameronnunez reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @cameronnunez, @knz, @rauchenstein, and @tbg)

___
*[-- commits, line 2 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MuBgEiA9tNFqTcUIvzO):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

What's up with this commit message? I think it's wrong and belongs to a different PR.
</blockquote></details>

Must have been some mistake during an interactive rebase/conflict resolution. Fixed.
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 40 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDSM-a7ck9qBs9QUzk:-Mu8Tudj7CHkSPI0PGl8:bk1mhvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L40)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: unclear why this isn't just the shorter `c.Put`.
</blockquote></details>

It's deprecated.
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 85 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDSi9s25pd9vk7_aOQ:-MuBuqvQBP5giZrKLlA9:b-10kn3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L85)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

How can this work? I saw earlier that you're giving this test a 30 minute timeout only.
</blockquote></details>

Fixed. The duration var is now specifically to be used here, and not the test timeout. 
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 86 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDT1vZCS3MZ0NMl02R:-MuBueReBibN1UtZAz_n:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L86)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Please use a `c.NewMonitor`, also you never `.Wait()` on this errgroup, so what's the point?
</blockquote></details>

Fixed.
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 91 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDSwTtExlBa-WhxZ1a:-MuBugAB-WN0UoCOmU0d:b-b4g5ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L91)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The numNodes-4 is confusing to me, mind adding a comment?
</blockquote></details>

Overlooked this, it was a mistake. Fixed.
___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 101 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDT6PF7RLQtWqdGrie:-MuBujTXChugEbrAWb9M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/slow_drain.go#L101)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ditto about monitor
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/tests/acceptance.go, line 71 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MtDS3jkEH7mMHUOmH2N:-Mu8Tq3T14WzzbUfdjKe:bn11kpu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f02cdef987d8d622b38ab8e3c5f1962158242d2f/pkg/cmd/roachtest/tests/acceptance.go#L71)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'm hesitant about adding this as an acceptance test - these run on *every* CI. Is that really what you want? I figured you'd add this as a free-standing test, since in the test it seems like you're running it for tens of minutes.
</blockquote></details>

Free-standing test is definitely the way to go :thumbsup: 


<!-- Sent from Reviewable.io -->

cameronnunez reviewed (commented):
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>
cli and server changes LGTM.
A question about the roachtest.

The commit message (as reported by reviewable) is weird.

When you edit that commit message, please explain in there (not in the release note) that this only covers half of the story, because this logging only tells us whether transfers are attempted or not, and they don't tell us why they fail. You can then go on to explain that the precise cause is decided on another node, the one that's driving the transfer (that's not the same node as the one that's decommissioning), and we don't yet have logging/tracing infra to gather that data in one place properly.

I recall there was a discussion with Andrei about this somewhere? Can you find it? And then link a URL in the commit message.
</blockquote></details>

@knz You mention decommissioning but this applies to draining as well right? Also, I think I may have found it but it seems like it concerns decommissioning in particular, not draining. Is it issue #69245?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-:-MuBzWwn9u0cy5EjhSyl:b-q4l8nf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @knz, @rauchenstein, and @tbg)

___
*[pkg/cmd/roachtest/tests/slow_drain.go, line 132 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MuBgYz88weJCUGFGExe:-MuBxTeX8ZRMjhOSHb8H:bugcvrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/81379d2270a553947308d4a2bc680ca506ed04be/pkg/cmd/roachtest/tests/slow_drain.go#L132)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I feel stupid, I'm not seeing it: can you point me to where verboseStoreLogRe is defined?
</blockquote></details>

> ```go
> func runSlowDrain(ctx context.Context, t test.Test, c cluster.Cluster, duration time.Duration) {
>	const (
>		numNodes          = 6
>		pinnedNodeID      = 1
>		replicationFactor = 5
>	)
>
>	var verboseStoreLogRe = regexp.MustCompile("failed to transfer lease")
> ```
</details>


Near the function header
___
*[pkg/server/serverpb/admin.proto, line 408 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-MuBfnBL4E_wvkN90Ndp:-MuBxutN4etyIozMZDXo:bs7wumb) ([raw file](https://github.com/cockroachdb/cockroach/blob/81379d2270a553947308d4a2bc680ca506ed04be/pkg/server/serverpb/admin.proto#L408)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: "during the range lease drain phase."
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
LGTM

> You mention decommissioning but this applies to draining as well right?

Yes both: we need participation from another node for extra logging for _both_ lease transfers (drain) and replica transfers (deco).

>  I think I may have found it but it seems like it concerns decommissioning in particular

It's not this issue (and we need to dig to find the right issue, but we don't need to do this right now).

However, the topic remains open, and so I suggest to merge this but leave the linked issue #65659 open for now. Let's have a chat at team meeting to discuss the bit that's missing (which will be in a later/different PR).

> Reviewed 1 of 3 files at r13, 2 of 2 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-:-MuC7LZzDluZo_cAOaqo:b-w9svje)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @knz, @rauchenstein, and @tbg)



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
The issue I was referring to is this one: #74158. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68488#-:-MuC7xZYDQCUngz3bOse:b-zdxcau)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @knz, @rauchenstein, and @tbg)



<!-- Sent from Reviewable.io -->

cameronnunez commented:
@knz @tbg @rauchenstein thanks!

cameronnunez commented:
bors r=knz
